lib/protocol/encryption.go

```diff
@@ -0,0 +1,521 @@
+	}
+
+	// Encrypt / adjust the request parameters.
+
+	origSize := size
+	if size < minPaddedSize {
+		// Make a request for minPaddedSize data instead of the smaller
+		// block. We'll chop of the extra data later.
+		size = minPaddedSize
+	}
+	encName := encryptName(name, folderKey)
+	encOffset := offset + int64(blockNo*blockOverhead)
+	encSize := size + blockOverhead
+
+	// Perform that request, getting back and encrypted block.
+
+	bs, err := e.Connection.Request(ctx, folder, encName, blockNo, encOffset, encSize, nil, 0, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the decrypted block (or an error if it fails decryption)
+
+	fileKey := FileKey(name, folderKey)
+	bs, err = DecryptBytes(bs, fileKey)
+	if err != nil {
+		return nil, err
+	}
+	return bs[:origSize], nil
+}
+
+func (e encryptedConnection) DownloadProgress(ctx context.Context, folder string, updates []FileDownloadProgressUpdate) {
+	if _, ok := e.folderKeys[folder]; !ok {
+		e.Connection.DownloadProgress(ctx, folder, updates)
+	}
+
+	// No need to send these
+}
+
+func (e encryptedConnection) ClusterConfig(config ClusterConfig) {
+	// TODO: Filter/clean the outgoing ClusterConfig?
+	e.Connection.ClusterConfig(config)
+}
+
+func encryptFileInfos(files []FileInfo, folderKey *[keySize]byte) {
+	for i, fi := range files {
+		files[i] = encryptFileInfo(fi, folderKey)
+	}
+}
+
+// encryptFileInfo encrypts a FileInfo and wraps it into a new fake FileInfo
+// with an encrypted name.
+func encryptFileInfo(fi FileInfo, folderKey *[keySize]byte) FileInfo {
+	fileKey := FileKey(fi.Name, folderKey)
+
+	// The entire FileInfo is encrypted with a random nonce, and concatenated
+	// with that nonce.
+
+	bs, err := proto.Marshal(&fi)
+	if err != nil {
+		panic("impossible serialization mishap: " + err.Error())
+	}
+	encryptedFI := encryptBytes(bs, fileKey)
+
+	// The vector is set to something that is higher than any other version sent
+	// previously, assuming people's clocks are correct. We do this because
+	// there is no way for the insecure device on the other end to do proper
+	// conflict resolution, so they will simply accept and keep whatever is the
+	// latest version they see. The secure devices will decrypt the real
+	// FileInfo, see the real Version, and act appropriately regardless of what
+	// this fake version happens to be.
+
+	version := Vector{
+		Counters: []Counter{
+			{
+				ID:    1,
+				Value: uint64(time.Now().UnixNano()),
+			},
+		},
+	}
+
+	// Construct the fake block list. Each block will be blockOverhead bytes
+	// larger than the corresponding real one and have an encrypted hash.
+	// Very small blocks will be padded upwards to minPaddedSize.
+	//
+	// The encrypted hash becomes just a "token" for the data -- it doesn't
+	// help verifying it, but it lets the encrypted device do block level
+	// diffs and data reuse properly when it gets a new version of a file.
```

Comment:
Not sure why but this triggers my security spider senses, because I know things like DARE, every time the file changes, everything in the files changes, so no reuse is possible.

I guess it's a compromise we're making, and that might be fine, but the nonce of the data changes, changing the encrypted payload, yet the fake hash stays the same, making me doubt the usefulness of the nonce, as effectively you know two differently looking encrypted payloads represent the same data, effectively providing an oracle.

Reply:
I don't follow: You are saying it's possible for an untrusted node to know that two blocks, differing in their encrypted form due to different nonces, have the same payload, correct? How would they know that?

Reply:
The encrypted hash is encrypted deterministically for reuse purposes as the comment says. So I assume the encrypted hash would give away that it's the same plaintext data (as otherwise how can you reuse anything?)

Reply:
When you request a data block it gets encrypted using a random nonce. You can request a block as many times as you like, getting a new random nonce and hence a new "variant" of the encrypted block. You can of course assume that the plaintext is the same every time you request it, unless the file changed on disk. I don't think this is a problem, this is generally how nonces work? The fact that we have a token of sorts that also says the plaintext didn't change is kind of secondary.

Reply:
And the comment is not referring to block reuse between different files, only from the old files. Block hashes are encrypted with the file key, i.e. the filename is used as a nonce and thus the same block in different files have different encrypted hashes.

Reply:
So I am not sure that's ok from some previous discussions on the subject on DARE/sio with @aead because:

1. If I am able to inject a file into the system somehow, by seeing that file propagate, I can verify if a file with similar content was there before because the deterministic hash will match. I can also verify that two files are the same even if they are at different locations, encrypted with different nonces etc.
2. Because I can ask for the same thing over and over again, and I'll get the same plaintext encrypted with the same key just a different nonce. I am not sure that's ok. Sure, 24 bytes is not easy to bruteforce but still. 

I'd be nice for @aead to provide an opinion.
I know when I discussed this in the past any prospect of block reuse looked unlikely.

Reply:
1. I still don't get how you can verify anything after injecting a file: The hash won't match with other files that have the same blocks, because:

```
encryptedBlockHash = AES-SIV(blockHash, fileKey)

fileKey = HKDF(SHA256, folderKey+filename, salt: "syncthing", info: nil)
```

2. Related question: Is there any gain in using a nonce/non-deterministic encryption at all then? As you state, the untrusted device anyway has the information to match encrypted data of the same plaintext (by knowing the deterministic encrypted hash). That's also true for encrypted file infos (knowing the deterministic encrypted filename). Even if that were fine in terms of security of the encryption (I don't know about that), as far as I understand there's no point in using a nonce in this scenario. Any in-transit attack scenarios don't apply to us, as transit encryption is covered by another layer. So we might as well use AES-SIV everywhere, which is designed for this scenario (no nonce, deterministic encryption).

Reply:
Hi, I don't think I understand the full scope / entire threat model. However, when a stream is split into fragments / blocks and each encrypted block is self-contained then there is a fundamental problem for data integrity:

Assume a file consisting of: `B0 || B1 || B2`. The encrypted file be: `C0 || C1 || C2`
Now, the file gets updated to `B0 || B1' || B2` (`B1 != B1'`). This leads to an encrypted file `C0 || C1' || C2` (`C1 != C1'`)

Now, a receiver has no way to determine what's the latest version of the file -  `C0 || C1 || C2` or `C0 || C1' || C2`
Therefore a malicious node can replace `C1'` with `C1` at any point in time and can re-distribute old content. If a node updates the file whenever `encryptedBlockHash` changes it would accept "outdated" content from a malicious node.
 
Also, the malicious node learns which parts of the file were changed. Depending on the file type this may allow plaintext recovery / usage analysis.

Reply:
> Now, a receiver has no way to determine what's the latest version of the file

Why do you think so? The file metadata is encrypted and signed and sent in the usual way. Certainly a malicious device can simply refuse to pass on updated metadata, but it can't create a new metadata entry that refers to old data.

> the malicious node learns which parts of the file were changed

This is indeed the intentional tradeoff. The alternative is that the entire file needs to be send on any change, and that behavior is far enough away from Syncthing that there's really no point any more, I think. At that point you can simply use an encryption program of your choice and SCP instead.

Reply:
> Why do you think so? The file metadata is encrypted and signed and sent in the usual way. Certainly a malicious device can simply refuse to pass on updated metadata, but it can't create a new metadata entry that refers to old data.

Not sure whether I'm missing something here but what prevents the following flow:
```
 - Node_s encrypts file: C0 || C1 || C2  Enc(B0) || Enc(B1) || Enc(B2) and distributes C0, C1 and C2
 - Node_x receives  C0, C1 and C2 from Node_s
 - Node_r receives C0, C1 and C2 from Node_x and reconstructs B0 || B1 || B2 = Dec(C0) || Dec(C1) || Dec(C2)
 - Node_s updates and encrypts: C1' = Enc(B1') (B1 != B1')  and distributes C1
 - Node_x receives C1'  from Node_s
 - Node_x receives C1' from Node_x and updates  B1 as B1' = Dec(C1')

 - Node_m sends  C1 to Node_x
 - Node_x receives C1 and has no way to determine whether it's a "new" block or an "old" one.
 - Node_r receives C1 and only rejects it if there is some "version downgrade protection" in place that rejects a block
   if its version (or other mono. increasing info) is older then the current info.
``` 
An adversary can replay old information to all peers that cannot apply the downgrade protection due to encryption.

Reply:
I misunderstood that the blocks are not distributed themselves on an update event. However, if I've understood the description correctly the same problem still exists. Whether it's the actual data or the metadata - at the point when there are multiple valid "versions" of a data structure (`B1'` being the successor of `B1` - or `F'` being the fileinfo successor of `F`) an attacker can replay the old version.
Whether an attacker replays an old ciphertext block or an old ciphertext metadata structure - which the receiver then uses to fetch the ciphertext block - is basically the same. If there is a downgrade-protection in place then a receiver will not accept any plaintext (block) with a lower version number, and therefore, reject the replayed data.

This boils down to: If a node receives multiple encrypted file infos (one for each file update) such that it fetches , decrypts **and** accepts:
 1. `... || B1`
 2. `... || B1'`
 3. `... || B1`
then an attacker can cause re-distribution of old content.

> The question right now is whether it is a problem, that an adversary can request the same blocks again and again, and we'll send him the same plaintext encrypted with chacha and a different nonce every time. And given the adversary knows that it's the same plaintext, is there any point/advantage in using chacha (with nonce, non-deterministic) instead of AES-SIV (deterministic), which we use in other places where the result must be deterministic for the protocol to work.

From a conceptual point of view there should be no difference. AES-GCM-SIV (as nonce-misuse-resistant AEAD) achieves the same security properties as ChaCha20-Poly1305 as long as a nonce is used and should reveal whether two ciphertexts correspond to the same plaintext if no nonce is used.
(AES-GCM-SIV is probabilistic if a nonce / random AAD is specified and deterministic otherwise - See e.g.: https://tools.ietf.org/html/rfc8452#section-1)

Reply:
> an attacker can replay the old version

They can, but the version vector identifies it as an old version. It will not cause the recipient to download the old version, it will cause them to consider the sender out of date.

Reply:
> They can, but the version vector identifies it as an old version. It will not cause the recipient to download the old version, it will cause them to consider the sender out of date.

:+1: 

> Here I'm not sure. :) I started with XChaCha20-Poly1305 because it seems like a modern construction suited for the purpose. AES-SIV came into the picture later for the specific cases where we need the property that a given plaintext corresponds to a certain ciphertext. It's possible we could use the latter for the former, but I saw no reason to switch. Is there a reason to switch?

Well, based on the comments above an adversary can request a encrypted block at any point in time. So, you provide an Oracle to the attacker. Since you use `XChaCha20-Poly1305` that uses 192 bit nonces (128 bit for HChaCha20 KDF and 96 bit for ChaCha20Poly1305) it is practically infeasible for an attacker to request blocks until nonce reuse becomes "likely". The attack would be:
```
1. Request C0 = Enc(B0)
2. Wait for B0 to change: B0' = B0 
3. Request C0' = Enc(B0')
4. Do some basic analysis to find out whether the nonce was reused - e.g. xor C0 and C0'
5. If the nonce was reused => goto 3. 
```
An attacker has to request a block at least `2^64` times to have a non-negligible chance of nonce reuse.
***
However, as an alternative you use AES-GCM-SIV (with or without a nonce). The advantage of AES-GCM-SIV is that
even though the nonce is 96 bits long and an attacker probably can trigger nonce reuse AES-GCM-SIV would only reveal
whether the block is equal to another block in the past in such a case.
However, the general considerations w.r.t. `(X)ChaCha20-Poly1305 vs AES-GCM` still applies - e.g. AES usually requires special hardware instructions for a secure & fast implementation.

So, I don't see a strong reason to use AES-GCM-SIV over XChaCha20-Poly1305 here. However, it is important that the nonce of non-deterministic scheme that does not provide misuse-resistance has to be sufficiently large. For instance, I think XChaCha20-Poly1305 would be fine while ChaCha20-Poly1305 (with 96 bit nonces) would not...